scipy/interpolate/_cubic.py

```diff
@@ -330,3 +335,163 @@ def from_spline(cls, tck, extrapolate=None):
     def from_bernstein_basis(cls, bp, extrapolate=None):
         raise NotImplementedError("This method does not make sense for "
                                   "an Akima interpolator.")
+
+
+class CubicSpline(PPoly):
+    """Cubic spline data interpolator.
+
+    Interpolate data with a piecewise cubic polynomial which is twice
+    continuously differentiable. The result is represented as a `PPoly`
+    instance with breakpoints matching the given data.
+
+    Parameters
+    ----------
+    x : array_like, shape (n,)
+        1-d array containing increasing values of the independent variable.
+    y : array_like
+        Array containing values of the dependent variable. It can have
+        arbitrary number of dimensions, but the length along `axis` (see below)
+        must match the length of `x`.
+    axis : int, optional
+        Axis along which `y` is assumed to be varying. Meaning that for
+        ``x[i]`` the corresponding values are ``np.take(y, i, axis=axis)``.
+        Default is 0.
+    extrapolate : bool, optional
+        Whether to extrapolate to ouf-of-bounds points based on first and last
+        intervals, or to return NaNs. Default is True.
+
+    Attributes
+    ----------
+    x : ndarray, shape (n,)
+        Breakpoints. The same `x` which was passed to the constructor.
+    c : ndarray, shape (4, n-1, ...)
+        Coefficients of the polynomials on each segment. The trailing
+        dimensions match the dimensions of `y`, excluding `axis`. For example,
+        if `y` is 1-d, then ``c[k, i]`` is a coefficient for
+        ``(x-x[i])**(3-k)`` on the segment between ``x[i]`` and ``x[i+1]``.
+    axis : int
+        Interpolation axis. The same `axis` which was passed to the
+        constructor.
+
+    Methods
+    -------
+    __call__
+    derivative
+    antiderivative
+    integrate
+    roots
+
+    See Also
+    --------
+    Akima1DInterpolator
+    PchipInterpolator
+
+    Notes
+    -----
+    .. versionadded:: 0.18.0
+
+    There are two additional equations required to determine all coefficients
+    of polynomials on each segment. They are formed from the conditions, that
+    at the first and the last interior knots the third derivative is
+    continuous. It essentially means that the first and the second segments
+    (on both ends) are described by the same cubic polynomial. These conditions
+    are called "not-a-knot" and are known to give better interpolation accuracy
+    when nothing is known about end point derivatives [1]_.
+
+    References
+    ----------
+    .. [1] de Boor, C., "A Practical Guide to Splines, Springer-Verlag", 1978.
+    """
+    def __init__(self, x, y, axis=0, extrapolate=True):
+        x, y = map(np.asarray, (x, y))
```

Comment:
Some more drudgery is needed in case x or y are integer or complex, see Akima1DInterpolator or Pchip. And it needs smoke tests with both x and y intege and complex y.